langkit/response_consistency.py

```diff
@@ -0,0 +1,241 @@
+
+            Passage: {response}
+
+            Is the passage supported by the context above?
+            Answer between: Accurate, Minor Inaccurate, Major Inaccurate
+            """
+            result: ChatLog = Conversation(self.consistency_checker_llm).send_prompt(
+                prompt
+            )
+            llm_score = self.convert_score(result.response)
+            llm_scores.append(llm_score)
+            total_tokens += result.total_tokens
+        final_score = sum(llm_scores) / len(llm_scores)
+        return (final_score, total_tokens)
+
+    def consistency_check_with_samples(self, response, additional_samples):
+        additional_samples = [ChatLog("", sample) for sample in additional_samples]
+        llm_score, tokens_usage = self.llm_consistency_check(
+            response, additional_samples
+        )
+
+        semantic_score = self.semantic_similarity(response, additional_samples)
+        consistency_result = ConsistencyResult(
+            llm_score=llm_score,
+            semantic_score=semantic_score,
+            final_score=(llm_score + semantic_score) / 2,
+            total_tokens=tokens_usage,
+            samples=[sample.response for sample in additional_samples],
+            response=response,
+        )
+        return consistency_result
+
+    def consistency_check(
+        self, prompt: str, response: Optional[str] = None
+    ) -> ConsistencyResult:
+        total_tokens = 0
+
+        if not response:
+            result: ChatLog = Conversation(self.llm).send_prompt(prompt)
+            if result.total_tokens:
+                total_tokens += result.total_tokens
+            if result.response:
+                response = result.response
+            else:
+                raise Exception(f"Error generating response: {result.errors}")
+
+        additional_samples: List[ChatLog] = self.get_samples(prompt)
+        total_tokens += sum(
+            [
+                sample.total_tokens if sample.total_tokens else 0
+                for sample in additional_samples
+            ]
+        )
+
+        llm_score, tokens_usage = self.llm_consistency_check(
+            response, additional_samples
+        )
+        total_tokens += tokens_usage
+
+        semantic_score = self.semantic_similarity(response, additional_samples)
+        consistency_result = ConsistencyResult(
+            llm_score=llm_score,
+            semantic_score=semantic_score,
+            final_score=(llm_score + semantic_score) / 2,
+            total_tokens=total_tokens,
+            samples=[sample.response for sample in additional_samples],
+            response=response,
+        )
+        return consistency_result
+
+
+checker: Optional[ConsistencyCheker] = None
+
+
+def init(llm: LLMInvocationParams, num_samples=1):
+    global checker
+    checker = ConsistencyCheker(llm, num_samples, embeddings_encoder)
+
+
+@register_dataset_udf([_prompt, _response], f"{_response}.consistency")
```

Comment:
Maybe we can rename the UDF function and the registration name as `response.hallucination` or something like that?